app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,197 @@
-    @object, workers_hierarchy, @background_destroy_method = job.arguments
-    id = "Hierarchy-#{object.class.name}-#{object.id}"
-    @caller_worker_hierarchy = Array(workers_hierarchy) + [id]
-    info "Starting #{job.class}#perform with the hierarchy of workers: #{caller_worker_hierarchy}"
+  rescue_from(DoNotRetryError) do |exception|
+    # report error and skip retries
+    System::ErrorReporting.report_error(exception)
   end
 
-  after_perform do |job|
-    info "Finished #{job.class}#perform with the hierarchy of workers: #{caller_worker_hierarchy}"
+  # we need this only for compatibility to process already enqueued jobs after upgrade
+  rescue_from(ActiveJob::DeserializationError) do |exception|
+    System::ErrorReporting.report_error(exception)
   end
 
-  def perform(_object, _caller_worker_hierarchy = [], _background_destroy_method = 'destroy')
-    build_batch
-  end
+  queue_as :deletion
 
-  def on_success(_, options)
-    on_finish('on_success', options)
-  end
+  # until_executed seems to rely on #after_perform which is skipped on failure so not sure whether retries will work.
+  # additionally if we want to reschedule ourselves in #after_perform, it is not nice to rely on flaky order
+  # see https://github.com/3scale/activejob-uniqueness/blob/main/lib/active_job/uniqueness/strategies/until_executed.rb
+  unique :until_and_while_executing, lock_ttl: 6.hours
 
-  def on_complete(_, options)
-    on_finish('on_complete', options)
-  end
+  # better limit by available  `delete` executors
+  # sidekiq_throttle concurrency: { limit: 10 }
 
-  def on_finish(method_name, options)
-    workers_hierarchy = options['caller_worker_hierarchy']
-    info "Starting DeleteObjectHierarchyWorker##{method_name} with the hierarchy of workers: #{workers_hierarchy}"
-    object = GlobalID::Locator.locate(options['object_global_id'])
-    background_destroy_method = @background_destroy_method.presence || 'destroy'
-    DeletePlainObjectWorker.perform_later(object, workers_hierarchy, background_destroy_method)
-    info "Finished DeleteObjectHierarchyWorker##{method_name} with the hierarchy of workers: #{workers_hierarchy}"
-  rescue ActiveRecord::RecordNotFound => exception
-    info "DeleteObjectHierarchyWorker##{method_name} raised #{exception.class} with message #{exception.message}"
+  before_perform do |job|
+    info "Starting #{job.class}#perform with the hierarchy of workers: #{job.arguments}"
   end
 
-  protected
-
-  delegate :info, to: 'Rails.logger'
+  after_perform do |job|
+    if @remaining_hierarchy.present?
+      self.class.perform_later(*@remaining_hierarchy)
+      msg = " iteration of"
+    end
+    info "Finished#{msg} #{job.class}#perform with the hierarchy of workers: #{job.arguments}"
+  end
 
-  attr_reader :object, :caller_worker_hierarchy
+  class << self
+    # convenience method to schedule deleting an active record object
+    def delete_later(ar_object)
+      perform_later *hierarchy_entries_for(ar_object)
```

Comment:
```suggestion
      perform_later(*hierarchy_entries_for(ar_object))
```
![Screenshot From 2025-03-31 12-36-28](https://github.com/user-attachments/assets/f273cdbc-0662-4177-b62f-b9eaf6de357f)